src/main/java/com/minecolonies/coremod/colony/workorders/WorkOrderBuildDecoration.java

```diff
@@ -182,47 +182,57 @@ public void attemptToFulfill(@NotNull final Colony colony)
 
             if (!job.hasWorkOrder() && canBuild(citizen))
             {
-                job.setWorkOrder(this);
-                this.setClaimedBy(citizen);
-                return;
+                final double distance = citizen.getWorkBuilding().getID().distanceSq(this.buildingLocation);
+                if(claimedBy == null || distance < distanceToBuilder)
+                {
+                    claimedBy = citizen;
+                    distanceToBuilder = distance;
+                }
             }
         }
 
+        if(claimedBy != null)
+        {
+            final JobBuilder job = claimedBy.getJob(JobBuilder.class);
+            job.setWorkOrder(this);
+            this.setClaimedBy(claimedBy);
+            return;
+        }
+
         sendBuilderMessage(colony, hasBuilder, sendMessage);
     }
 
     /**
      * send a message from the builder.
-     *
+     * <p>
      * Suppressing Sonar Rule squid:S1172
      * This rule does "Unused method parameters should be removed"
      * But in this case extending class may need to use the sendMessage parameter
-     * @param colony which the work order belong to
-     * @param hasBuilder true if we have a builder for this work order
+     *
+     * @param colony      which the work order belong to
+     * @param hasBuilder  true if we have a builder for this work order
      * @param sendMessage true if we need to send the message
      */
     @SuppressWarnings("squid:S1172")
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Suppressing the 'squid:S1172' warning is not allowed [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1309)